### ISSUE TITLE
Potential loss of bandwidth through asyncio transport sendto

### ISSUE THREAD

**Issue by qstokkink** [Tribler/py-ipv8#980]
In #975 I encountered the issue that using the asyncio UDP transport to send data would block for several minutes, whereas using the raw socket to send data would take only in the order of hundreds of milliseconds.

It just dawned on me that this is also what we do in the `UDPEndpoint` class:

https://github.com/Tribler/py-ipv8/blob/271d62b26eee88839db4630abcca14e67fb92d36/ipv8/messaging/interfaces/udp/endpoint.py#L46

If the same issue holds, we are seemingly missing out on a lot of throughput. We should investigate adopting the send from the `BroadcastBootstrapEndpoint`:

https://github.com/Tribler/py-ipv8/blob/145d83cefe7ac4b94ef787d61547efc5a47e3b38/ipv8/bootstrapping/udpbroadcast/bootstrapper.py#L46

Even though this may seem cut and dried, there is also a chance that this additional slowdown for illegal addresses is doing something useful for well-known addresses. I.e., long term benefit over short term benefit.

We won't know until we test.

**Comment by qstokkink** [comment]
I had some fun with `iperf3` today to measure our raw socket throughput (the `UDPEndpoint` with a single listener attached). I also tested sending to unknown addresses. These are the results:

```
== transport.sendto (current) ==
Megabytes per second: 359.71 MBps
Send time to unknown: 15.86 microseconds

== socket.sendto (proposed) ==
Megabytes per second: 375.94 MBps
Send time to unknown: 12.99 microseconds
```

(Sample sizes: 10 seconds of 25 GBit/s send for the throughput test and 1000 fictional-but-equal addresses for the send time)

Observations:

 1. `socket.sendto` has higher throughput and has lower delay when sending to unknown addresses.
 2. I could not replicate the huge delay I uncurred with the broadcast socket.
 
In conclusion, it would be _a little bit_ better to use `socket.sendto` instead of `transport.sendto`. But, we're in no hurry to implement this.

**Comment by synctext** [comment]
awesome! my multi-year wish is a performance test we run each PR. Showing exactly what you measured today of raw sockwts, UDP via IPv8, UDP in a single community receiver, UDP through Libtorrent, and ending with our 6MBps-ish exit node experience for end-users.

**Comment by qstokkink** [comment]
@synctext [Disclaimer: another undocumented lesson from the past.] At some point we did have a PR test for endpoint performance, enforcing a lack of significant performance degradation. However, I removed it for these three reasons:

 1. In order to perform a reliable performance test, the test machine has to be _completely_ isolated and inert. Any parallel experiment/test run/OS update/other machine in the LAN/printers/etc. that uses the machine's network messes up the measurements. In practice, enforcing a lack of significant performance degradation was just forcing us to retest PRs until they finally passed by chance.
 2. Any new feature slows down the packet handling. As asyncio Python is single-threaded-ish (I won't **G**o **I**nto detai**L**), every line of code you add takes away processing time from the socket and thus causes lower throughput. In other words, for every feature that is added, we would have to make some arbitrary decision whether "feature X is worth a throughput drop of Y kbps". Even worse, most communities have complex logic that would claim main thread execution sporadically and/or under very specific conditions, making it insanely hard to (automatically) quantify the real throughput impact.
 3. Performance testing is slow. The unit tests complete in 10 seconds, a reliable performance test takes a few minutes.

These reasons are why I stopped running it _every PR_, but, knowing the performance impact of each component of IPv8 and Tribler is of course useful. Instead of doing this every PR, I'd propose sticking to periodic (daily?) tests instead.

**Comment by egbertbouman** [comment]
I also did some testing with `sendto` last week for the `TunnelCommunity`. I'm getting a ~15% bump in throughput while testing this on localhost with the `ProactorEventLoop`, but with the `SelectorEventLoop` (the one we use in Tribler)  I'm noticing no real improvement. Of course, I'm not just measuring `sendto`, but all other IPv8/tunnel logic as well.

Note that if we start bypassing the `asyncio` transport, we're no longer buffering the packets (at least while using the `SelectorEventLoop`), but we just drop them.

**Comment by qstokkink** [comment]
Closing this for the following reasons:

Firstly, after some digging, it seems that the lockup on the Bootstrapper was specific to my machine (I still don't know why it is happening). Whatever is happening, it does not seem to occur under normal endpoint use.

Secondly, the main reason to speed up the socket is to gain TunnelCommunity performance. @egbertbouman showed there is little to no performance gain, under increased risk of packet drop: no positive and an extra negative.
